tools/generator/src/Generator/XCSchemeInfo+LaunchActionInfo.swift

```diff
@@ -0,0 +1,127 @@
+
+extension XCSchemeInfo.LaunchActionInfo {
+    /// Create a copy of the launch action info with host in the target info resolved.
+    init?(
+        resolveHostsFor launchActionInfo: XCSchemeInfo.LaunchActionInfo?,
+        topLevelTargetInfos: [XCSchemeInfo.TargetInfo]
+    ) throws {
+        guard let original = launchActionInfo else {
+          return nil
+        }
+        try self.init(
+            buildConfigurationName: original.buildConfigurationName,
+            targetInfo: .init(
+                resolveHostFor: original.targetInfo,
+                topLevelTargetInfos: topLevelTargetInfos
+            ),
+            args: original.args,
+            env: original.env,
+            workingDirectory: original.workingDirectory
+        )
+    }
+}
+
+// MARK: runnable
+
+extension XCSchemeInfo.LaunchActionInfo {
+    var runnable: XCScheme.Runnable? {
+        // To produce a similar result to the current generation code, we need to have a
+        // LaunchActionInfo, but we do not want to set the runnable.
+        if targetInfo.pbxTarget.isTestable {
+            return nil
+        }
+        if targetInfo.isWidgetKitExtension {
+            return XCScheme.RemoteRunnable(
+                buildableReference: targetInfo.buildableReference,
+                bundleIdentifier: "com.apple.springboard",
+                runnableDebuggingMode: "2"
+            )
+        } else {
+            // If targeting a device for a Watch App, Xcode modifies the scheme
+            // to use a `RemoteRunnable`. It does this automatically though, so
+            // we don't have to account for it
+            return XCScheme.BuildableProductRunnable(
+                buildableReference: targetInfo.buildableReference
+            )
+        }
+    }
+}
+
+// MARK: askForAppToLaunch
+
+extension XCSchemeInfo.LaunchActionInfo {
+    var askForAppToLaunch: Bool {
+        return targetInfo.isWidgetKitExtension
+    }
+}
+
+// MARK: macroExpansion
+
+extension XCSchemeInfo.LaunchActionInfo {
+    var macroExpansion: XCScheme.BuildableReference? {
+        get throws {
+            if let hostBuildableReference = try targetInfo.selectedHostInfo?.buildableReference,
+                !targetInfo.productType.isWatchApplication
+            {
+                return hostBuildableReference
+            } else if targetInfo.pbxTarget.isTestable {
+                return targetInfo.buildableReference
+            }
+            return nil
+        }
+    }
+}
+
+// MARK: launcher
+
+extension XCSchemeInfo.LaunchActionInfo {
+    var launcher: String {
+        if targetInfo.productType.canUseDebugLauncher {
+            return XCScheme.defaultLauncher
+        }
+        return XCSchemeConstants.posixSpawnLauncher
+    }
+}
+
+// MARK: debugger
+
+extension XCSchemeInfo.LaunchActionInfo {
+    var debugger: String {
+        if targetInfo.productType.canUseDebugLauncher {
+            return XCScheme.defaultDebugger
+        }
+        return ""
```

Comment:
I like to use `guard`s for early exits:
```suggestion
        guard targetInfo.productType.canUseDebugLauncher else {
            return ""
        }
        return XCScheme.defaultDebugger
```